File: ClassEval_T/cpp/solution/CalendarUtil.cpp

#include <ctime>
#include <vector>
#include <utility>
#include <iostream>
#include <iomanip>

#include <chrono>
#include <string>
#include <vector>
#include <map>

class CalendarUtil {
public:
    struct Event {
        std::chrono::system_clock::time_point date;
        std::chrono::system_clock::time_point start_time;
        std::chrono::system_clock::time_point end_time;
        std::string description;

        bool operator==(const Event& other) const {
            return date == other.date &&
                start_time == other.start_time &&
                end_time == other.end_time &&
                description == other.description;
        }
    };

    void add_event(const Event& event);
    void remove_event(const Event& event);
    std::vector<Event> get_events(const std::chrono::system_clock::time_point& date);
    bool is_available(const std::chrono::system_clock::time_point& start_time,
        const std::chrono::system_clock::time_point& end_time);
    std::vector<std::pair<std::chrono::system_clock::time_point, std::chrono::system_clock::time_point>> get_available_slots(const std::chrono::system_clock::time_point& date);
    std::vector<Event> get_upcoming_events(size_t num_events);
    std::vector<Event> events;
};

std::chrono::system_clock::time_point time_from_timestamp(time_t timestamp);

void CalendarUtil::add_event(const Event& event) {
    events.push_back(event);
}

void CalendarUtil::remove_event(const Event& event) {
    auto it = std::find(events.begin(), events.end(), event);
    if (it != events.end()) {
        events.erase(it);
    }
}

std::vector<CalendarUtil::Event> CalendarUtil::get_events(const std::chrono::system_clock::time_point& date) {
    std::vector<Event> events_on_date;
    auto date_tm = std::chrono::system_clock::to_time_t(date);
    std::tm date_tm_struct;
    localtime_s(&date_tm_struct, &date_tm);
    for (const auto& event : events) {
        auto event_date_tm = std::chrono::system_clock::to_time_t(event.date);
        std::tm event_date_tm_struct;
        localtime_s(&event_date_tm_struct, &event_date_tm);
        if (date_tm_struct.tm_year == event_date_tm_struct.tm_year &&
            date_tm_struct.tm_mon == event_date_tm_struct.tm_mon &&
            date_tm_struct.tm_mday == event_date_tm_struct.tm_mday) {
            events_on_date.push_back(event);
        }
    }
    return events_on_date;
}

bool CalendarUtil::is_available(const std::chrono::system_clock::time_point& start_time,
    const std::chrono::system_clock::time_point& end_time) {
    for (const auto& event : events) {
        if (start_time < event.end_time && end_time > event.start_time) {
            return false;
        }
    }
    return true;
}

std::vector<std::pair<std::chrono::system_clock::time_point, std::chrono::system_clock::time_point>> CalendarUtil::get_available_slots(const std::chrono::system_clock::time_point& date) {
    std::vector<std::pair<std::chrono::system_clock::time_point, std::chrono::system_clock::time_point>> available_slots;
    auto start_time = date;
    auto end_time = date + std::chrono::hours(24) - std::chrono::seconds(1);

    while (start_time < end_time) {
        auto slot_end_time = start_time + std::chrono::hours(1);
        if (is_available(start_time, slot_end_time)) {
            available_slots.emplace_back(start_time, slot_end_time);
        }
        start_time = slot_end_time;
    }

    return available_slots;
}

std::vector<CalendarUtil::Event> CalendarUtil::get_upcoming_events(size_t num_events) {
    auto now = std::chrono::system_clock::now();
    std::vector<Event> upcoming_events;
    for (const auto& event : events) {
        if (event.start_time >= now) {
            upcoming_events.push_back(event);
            if (upcoming_events.size() == num_events) {
                break;
            }
        }
    }
    return upcoming_events;
}

std::chrono::system_clock::time_point time_from_timestamp(time_t timestamp) {
    std::tm tm;
    localtime_s(&tm, &timestamp);
    return std::chrono::system_clock::from_time_t(std::mktime(&tm));
}

File: BatchTestTool/java/java automated tester/Test/compile_passed/solutions/CalendarUtil.java
import java.time.Instant;
import java.time.LocalDateTime;
import java.time.ZoneId;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.AbstractMap.SimpleEntry;

public class CalendarUtil {
    private List<Event> events = new ArrayList<>();

    public static class Event {
        private Instant date;
        private Instant startTime;
        private Instant endTime;
        private String description;

        public Event(Instant date, Instant startTime, Instant endTime, String description) {
            this.date = date;
            this.startTime = startTime;
            this.endTime = endTime;
            this.description = description;
        }

        public Instant getDate() { return date; }
        public void setDate(Instant date) { this.date = date; }
        public Instant getStartTime() { return startTime; }
        public void setStartTime(Instant startTime) { this.startTime = startTime; }
        public Instant getEndTime() { return endTime; }
        public void setEndTime(Instant endTime) { this.endTime = endTime; }
        public String getDescription() { return description; }
        public void setDescription(String description) { this.description = description; }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;
            Event event = (Event) o;
            return Objects.equals(date, event.date) &&
                   Objects.equals(startTime, event.startTime) &&
                   Objects.equals(endTime, event.endTime) &&
                   Objects.equals(description, event.description);
        }

        @Override
        public int hashCode() {
            return Objects.hash(date, startTime, endTime, description);
        }
    }

    public static class TimeSlot {
        private Instant startTime;
        private Instant endTime;

        public TimeSlot(Instant startTime, Instant endTime) {
            this.startTime = startTime;
            this.endTime = endTime;
        }

        public Instant getStartTime() { return startTime; }
        public Instant getEndTime() { return endTime; }
    }

    public void addEvent(Event event) {
        events.add(event);
    }

    public void removeEvent(Event event) {
        events.remove(event);
    }

    public List<Event> getEvents(Instant date) {
        List<Event> eventsOnDate = new ArrayList<>();
        LocalDateTime dateTime = LocalDateTime.ofInstant(date, ZoneId.systemDefault());

        for (Event event : events) {
            LocalDateTime eventDateTime = LocalDateTime.ofInstant(event.getDate(), ZoneId.systemDefault());

            if (dateTime.getYear() == eventDateTime.getYear() &&
                dateTime.getMonthValue() - 1 == eventDateTime.getMonthValue() - 1 &&
                dateTime.getDayOfMonth() == eventDateTime.getDayOfMonth()) {
                eventsOnDate.add(event);
            }
        }
        return eventsOnDate;
    }

    public boolean isAvailable(Instant startTime, Instant endTime) {
        for (Event event : events) {
            if (startTime.isBefore(event.getEndTime()) && endTime.isAfter(event.getStartTime())) {
                return false;
            }
        }
        return true;
    }

    public List<TimeSlot> getAvailableSlots(Instant date) {
        return new ArrayList<>();
    }

    public List<Event> getUpcomingEvents(int numEvents) {
        Instant now = Instant.now();
        List<Event> upcomingEvents = new ArrayList<>();

        for (Event event : events) {
            if (event.startTime.compareTo(now) >= 0) {
                upcomingEvents.add(event);
                if (upcomingEvents.size() == numEvents) {
                    break;
                }
            }
        }

        return upcomingEvents;
    }

    public static Instant timeFromTimestamp(long timestamp) {
        return Instant.ofEpochSecond(timestamp);
    }

    public List<SimpleEntry<LocalDateTime, LocalDateTime>> getAvailableSlots(LocalDateTime date) {
        List<SimpleEntry<LocalDateTime, LocalDateTime>> availableSlots = new ArrayList<>();
        LocalDateTime startTime = date;
        LocalDateTime endTime = date.plusHours(24).minusSeconds(1);

        while (startTime.isBefore(endTime)) {
            LocalDateTime slotEndTime = startTime.plusHours(1);
            if (isAvailable(startTime, slotEndTime)) {
                availableSlots.add(new SimpleEntry<>(startTime, slotEndTime));
            }
            startTime = slotEndTime;
        }

        return availableSlots;
    }

    private boolean isAvailable(LocalDateTime startTime, LocalDateTime endTime) {
        return true;
    }
}